### PR TITLE
Alpha version of SSHD support with wolfSSH

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
This adds the alpha version of SSHD support to wolfSSH. Enabled with --enable-sshd when compiling with autotools and off by default. The core features are implemented and some additional features are still in progress. Some testing has been completed but additional testing for a production ready version of the code is in progress, so far Linux/QNX OS's have been tested, OSX has been built but password support not yet completed.

Once built with --enable-sshd an example of running the SSHD daemon would be the following:

```
sudo ./apps/wolfsshd/wolfsshd -h <host key> -f <sshd_config file> -p 22222
```

To see a help menu use `./apps/wolfsshd/wolfsshd -h`. The program can be run in the foreground using the option '-D'.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -210,7 +221,9 @@ AC_ARG_ENABLE([distro],
 AS_IF([test "x$ENABLED_DISTRO" = "xyes"],
       [ENABLED_ALL=yes; enable_shared=yes; enable_static=yes])
 AS_IF([test "x$ENABLED_ALL" = "xyes"],
-      [ENABLED_KEYGEN=yes; ENABLED_SCP=yes; ENABLED_SFTP=yes; ENABLED_FWD=yes; ENABLED_SHELL=yes; ENABLED_AGENT=yes])
+      [ENABLED_KEYGEN=yes; ENABLED_SCP=yes; ENABLED_SFTP=yes; ENABLED_FWD=yes; ENABLED_SHELL=yes; ENABLED_AGENT=yes; ENABLED_SSHD=yes])
+AS_IF([test "x$ENABLED_SSHD" = "xyes"],
+      [ENABLED_SHELL=yes;])
```
Comment (ejohnstown) [inline configure.ac:226]: Stray semicolon.

File: apps/wolfsshd/configuration.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/* configuration.h
+ *
+ * Copyright (C) 2014-2021 wolfSSL Inc.
```
Comment (ejohnstown) [inline apps/wolfsshd/configuration.h:3]: This is all over: Copyright 2022.

File: apps/wolfsshd/configuration.c [hunk 3]
```diff
@@ -0,0 +1,908 @@
...
+
+    /* No value, nothing to do */
+    if (!value || value[0] == '\0') {
+        return WS_BAD_ARGUMENT;
+    }
+
+    /* Ignore trailing whitespace */
+    ptr = value + WSTRLEN(value) - 1;
+    while (ptr != value) {
+        if (!isspace(*ptr)) {
```
Comment (ejohnstown) [inline apps/wolfsshd/configuration.c:448]: This is more a question. We use wrappers for malloc and free, and bunch of the string functions. Do we want to be mixing the macro versions and bare calls? Is this expected to be built for Linux only, or for other OSes?

File: apps/wolfsshd/auth.c [hunk 4]
```diff
@@ -0,0 +1,908 @@
...
+            wolfSSH_Log(WS_LOG_ERROR, "[SSHD] Login as root not permited");
+            return WOLFSSH_USERAUTH_FAILURE;
+        }
+    }
+
+    ret = DoCheckUser(usr, auth);
+    /* temporarily elevate permissions */
+    if (ret == WOLFSSH_USERAUTH_SUCCESS &&
+            wolfSSHD_AuthRaisePermissions(auth) != WS_SUCCESS) {
+        wolfSSH_Log(WS_LOG_ERROR, "[SSHD] Failure to raise permissions for auth");
```
Comment (ejohnstown) [inline apps/wolfsshd/auth.c:593]: Wrap string, and trailing whitespace.

File: apps/wolfsshd/wolfsshd.c [hunk 5]
```diff
@@ -0,0 +1,1017 @@
...
+    closelog();
+}
+
+#endif
+
+static void ShowUsage(void)
+{
+    printf("wolfsshd %s\n", LIBWOLFSSH_VERSION_STRING);
+    printf(" -?             display this help and exit\n");
+    printf(" -f <file name> Configuration file to use, default is /etc/ssh/sshd_config\n");
```
Comment (ejohnstown) [inline apps/wolfsshd/wolfsshd.c:146]: Wrap string.

File: apps/wolfsshd/wolfsshd.c [hunk 6]
```diff
@@ -0,0 +1,1017 @@
...
+                }
+
+                WFREE(data, NULL, 0);
+                wc_FreeDer(&der);
+            }
+        }
+    }
+
+    /* Load in host public key */
+//    {
```
Comment (ejohnstown) [inline apps/wolfsshd/wolfsshd.c:275]: Comment style. This is a good case for `#if 0`

**Review by ejohnstown — CHANGES_REQUESTED** [review]

File: apps/wolfsshd/configuration.c (hunk 3 above)
Comment (JacobBarthelmeh) [reply apps/wolfsshd/configuration.c:448]: if using the isspace we should abstract it to a macro, will fix it

**Review by ejohnstown — APPROVED** [review]

**Pull request merged**